pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/automaton/Automaton.java

```diff
@@ -0,0 +1,653 @@
+    s._transitionSet.add(new Transition(Character.MIN_VALUE, Character.MAX_VALUE, s));
+    for (State p : getStates()) {
+      int maxi = Character.MIN_VALUE;
+      for (Transition t : p.getSortedTransitions(false)) {
+        if (t._min > maxi) {
+          p._transitionSet.add(new Transition((char) maxi, (char) (t._min - 1), s));
+        }
+        if (t._max + 1 > maxi) {
+          maxi = t._max + 1;
+        }
+      }
+      if (maxi <= Character.MAX_VALUE) {
+        p._transitionSet.add(new Transition((char) maxi, Character.MAX_VALUE, s));
+      }
+    }
+  }
+
+  /**
+   * Reduces this automaton.
+   * An automaton is "reduced" by combining overlapping and adjacent edge intervals with same destination.
+   */
+  public void reduce() {
+    if (isSingleton()) {
+      return;
+    }
+    Set<State> states = getStates();
+    setStateNumbers(states);
+    for (State s : states) {
+      List<Transition> st = s.getSortedTransitions(true);
+      s.resetTransitions();
+      State p = null;
+      int min = -1;
+      int max = -1;
+      for (Transition t : st) {
+        if (p == t._to) {
+          if (t._min <= max + 1) {
+            if (t._max > max) {
+              max = t._max;
+            }
+          } else {
+            if (p != null) {
+              s._transitionSet.add(new Transition((char) min, (char) max, p));
+            }
+            min = t._min;
+            max = t._max;
+          }
+        } else {
+          if (p != null) {
+            s._transitionSet.add(new Transition((char) min, (char) max, p));
+          }
+          p = t._to;
+          min = t._min;
+          max = t._max;
+        }
+      }
+      if (p != null) {
+        s._transitionSet.add(new Transition((char) min, (char) max, p));
+      }
+    }
+    clearHashCode();
+  }
+
+  /**
+   * Returns sorted array of all interval start points.
+   */
+  char[] getStartPoints() {
+    //TODO: move to bitsets
+    Set<Character> pointset = new HashSet<Character>();
+    pointset.add(Character.MIN_VALUE);
+    for (State s : getStates()) {
+      for (Transition t : s._transitionSet) {
+        pointset.add(t._min);
+        if (t._max < Character.MAX_VALUE) {
+          pointset.add((char) (t._max + 1));
+        }
+      }
+    }
+    char[] points = new char[pointset.size()];
+    int n = 0;
+    for (Character m : pointset) {
+      points[n++] = m;
+    }
+    // Remove once move to bitsets
+    Arrays.sort(points);
+    return points;
+  }
+
+  private Set<State> getLiveStates(Set<State> states) {
+    HashMap<State, Set<State>> map = new HashMap<State, Set<State>>();
+    for (State s : states) {
+      map.put(s, new HashSet<>());
+    }
+    for (State s : states) {
+      for (Transition t : s._transitionSet) {
+        map.get(t._to).add(s);
+      }
+    }
+    Set<State> live = new HashSet<State>(getAcceptStates());
+    LinkedList<State> worklist = new LinkedList<State>(live);
+    while (worklist.size() > 0) {
```

Comment:
This has been commented on previously, this should use `!worklist.isEmpty()` everywhere, because `LinkedList.size()` traverses the list each time.